### PR TITLE
arrow-cpp: add GCS feature flag

### PR DESCRIPTION
**Pull request by cpcloud**
###### Motivation for this change

Add GCS support for arrow-cpp.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/#sec-conf-file))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/libraries/arrow-cpp/default.nix [hunk 1]
```diff
@@ -216,7 +216,7 @@ stdenv.mkDerivation rec {
     ''
       runHook preInstallCheck
 
-      ctest -L unittest -V \
+      ctest -L unittest \
```
Comment (veprbl) [inline pkgs/development/libraries/arrow-cpp/default.nix:219]: Does this affect verbosity of output for failures? Having verbose errors allows to gauge complexity of a build issue at a glance by just the Hydra build logs.

Comment (veprbl) [reply pkgs/development/libraries/arrow-cpp/default.nix:219]: Looking at OfBorg logs, they are still very long. I'm not sure what is the effect here.

Comment (cpcloud) [reply pkgs/development/libraries/arrow-cpp/default.nix:219]: Locally the logs are much shorter. Verbosity is unchanged for failures.

File: pkgs/development/libraries/arrow-cpp/default.nix [hunk 2]
```diff
@@ -34,6 +39,7 @@
 , enableShared ? !stdenv.hostPlatform.isStatic
 , enableFlight ? !stdenv.isDarwin # libnsl is not supported on darwin
 , enableS3 ? true
+, enableGcs ? true
```
Comment (veprbl) [inline pkgs/development/libraries/arrow-cpp/default.nix:42]:
```suggestion
, enableGcs ? !stdenv.isDarwin # google-cloud-cpp is not supported on darwin
```

Comment (cpcloud) [reply pkgs/development/libraries/arrow-cpp/default.nix:42]: Is this because of the transitive `libnsl` dependency (through `grpc`)?

**Comment by cpcloud** [comment]
Well, that's fun. It looks like we're hitting this https://github.com/boostorg/process/issues/55. In theory this should build if we use `boost170` on darwin. I'll add an assert so that people are aware of the need to upgrade to boost170 on darwin when s3 is enabled :grimacing:

**Pull request merged**